linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala

```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
   val ZipkinSampleHeader = "x-b3-sampled"
   val ZipkinFlagsHeader = "x-b3-flags"
 
+  /**
+   * The separate "b3" header in b3 single header format :
+   * b3={x-b3-traceid}-{x-b3-spanid}-{if x-b3-flags 'd' else x-b3-sampled}-{x-b3-parentspanid},
+   * where the last two fields are optional.
+   */
+  val ZipkinB3SingleHeader = "b3"
+
+  /**
+   * Handle b3 the same way as finagle handles b3, extract the traceid from "b3" header, write back "X-B3-*" multi
+   * headers, and let the flow continue as if it has received "x-b3-".
+   * @note Copied/adapted from finagle TraceInfo.scala
+   */
+  def convertB3SingleHeaderToMultiHeaders(headers: HeaderMap): Unit = {
+    if (headers.contains(ZipkinB3SingleHeader)) {
+      def handleSampled(headers: HeaderMap, value: String): Unit =
+        value match {
+          case "0" => {
+            headers.set(ZipkinSampleHeader, "0")
+            ()
+          }
+          case "d" => {
+            headers.set(ZipkinFlagsHeader, "1")
+            ()
+          }
+          case "1" => {
+            headers.set(ZipkinSampleHeader, "1")
+            ()
+          }
+          case _ => ()
+        }
+      def handleTraceAndSpanIds(headers: HeaderMap, a: String, b: String): Unit = {
+        headers.set(ZipkinTraceHeader, a)
+        headers.set(ZipkinSpanHeader, b)
+        ()
+      }
+
+      val _ = caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-").toList) match {
+        case Some(a) =>
```

Comment:
Since this is a side - effecting operation, you do not need to match on None really, as nothing useful is happening at that point.  Also,  matching on the size in this way is not very idiomatic. You can use cons matches I think. So it would look a bit like: 

```
      caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-").toList).foreach {
        case  Nil => ///
        case  a :: Nil =>
        case  a :: b :: Nil =>
        case  a :: b :: c :: Nil =>
      }
```

Reply:
No need to change to cons matches, since I've dropped the toList